---
title: Hook'larÄ±n KurallarÄ±
---

<Intro>
Hook'lar JavaScript fonksiyonlarÄ± kullanÄ±larak tanÄ±mlanÄ±r, ancak nerede Ã§aÄŸrÄ±labilecekleri konusunda kÄ±sÄ±tlamalarÄ± olan Ã¶zel bir yeniden kullanÄ±labilir UI mantÄ±ÄŸÄ± tÃ¼rÃ¼nÃ¼ temsil ederler.
</Intro>

<InlineToc />

---

##  Hook'larÄ± yalnÄ±zca en Ã¼st dÃ¼zeyde Ã§aÄŸÄ±rÄ±n {/*only-call-hooks-at-the-top-level*/}

Ä°simleri `use` ile baÅŸlayan fonksiyonlar React'te [*Hooks*](/reference/react) olarak adlandÄ±rÄ±lÄ±r.

**Hook'larÄ± dÃ¶ngÃ¼ler, koÅŸullar, iÃ§ iÃ§e fonksiyonlar veya `try`/`catch`/`finally` bloklarÄ± iÃ§inde Ã§aÄŸÄ±rmayÄ±n.** Bunun yerine, Hook'larÄ± her zaman React fonksiyonunuzun en Ã¼st seviyesinde, herhangi bir erken dÃ¶nÃ¼ÅŸten Ã¶nce kullanÄ±n. Hook'larÄ± yalnÄ±zca React bir fonksiyon bileÅŸenini iÅŸlerken Ã§aÄŸÄ±rabilirsiniz:

* âœ… BunlarÄ± bir [fonksiyon bileÅŸeninin](/learn/your-first-component) gÃ¶vdesinde en Ã¼st seviyede Ã§aÄŸÄ±rÄ±n.
* âœ… Call them at the top level in the body of a [custom Hook](/learn/reusing-logic-with-custom-hooks).

```js{2-3,8-9}
function Counter() {
  // âœ… Good: top-level in a function component
  const [count, setCount] = useState(0);
  // ...
}

function useWindowWidth() {
  // âœ… Good: top-level in a custom Hook
  const [width, setWidth] = useState(window.innerWidth);
  // ...
}
```

Itâ€™s **not** supported to call Hooks (functions starting with `use`) in any other cases, for example:

* ğŸ”´ Do not call Hooks inside conditions or loops.
* ğŸ”´ Do not call Hooks after a conditional `return` statement.
* ğŸ”´ Do not call Hooks in event handlers.
* ğŸ”´ Do not call Hooks in class components.
* ğŸ”´ Do not call Hooks inside functions passed to `useMemo`, `useReducer`, or `useEffect`.
* ğŸ”´ Do not call Hooks inside `try`/`catch`/`finally` blocks.

If you break these rules, you might see this error.

```js{3-4,11-12,20-21}
function Bad({ cond }) {
  if (cond) {
    // ğŸ”´ Bad: inside a condition (to fix, move it outside!)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad() {
  for (let i = 0; i < 10; i++) {
    // ğŸ”´ Bad: inside a loop (to fix, move it outside!)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad({ cond }) {
  if (cond) {
    return;
  }
  // ğŸ”´ Bad: after a conditional return (to fix, move it before the return!)
  const theme = useContext(ThemeContext);
  // ...
}

function Bad() {
  function handleClick() {
    // ğŸ”´ Bad: inside an event handler (to fix, move it outside!)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad() {
  const style = useMemo(() => {
    // ğŸ”´ Bad: inside useMemo (to fix, move it outside!)
    const theme = useContext(ThemeContext);
    return createStyle(theme);
  });
  // ...
}

class Bad extends React.Component {
  render() {
    // ğŸ”´ Bad: inside a class component (to fix, write a function component instead of a class!)
    useEffect(() => {})
    // ...
  }
}

function Bad() {
  try {
    // ğŸ”´ Bad: inside try/catch/finally block (to fix, move it outside!)
    const [x, setX] = useState(0);
  } catch {
    const [x, setX] = useState(1);
  }
}
```

You can use the [`eslint-plugin-react-hooks` plugin](https://www.npmjs.com/package/eslint-plugin-react-hooks) to catch these mistakes.

<Note>

[Custom Hooks](/learn/reusing-logic-with-custom-hooks) *may* call other Hooks (that's their whole purpose). This works because custom Hooks are also supposed to only be called while a function component is rendering.

</Note>

---

## Only call Hooks from React functions {/*only-call-hooks-from-react-functions*/}

Donâ€™t call Hooks from regular JavaScript functions. Instead, you can:

âœ… Call Hooks from React function components.
âœ… Call Hooks from [custom Hooks](/learn/reusing-logic-with-custom-hooks#extracting-your-own-custom-hook-from-a-component).

By following this rule, you ensure that all stateful logic in a component is clearly visible from its source code.

```js {2,5}
function FriendList() {
  const [onlineStatus, setOnlineStatus] = useOnlineStatus(); // âœ…
}

function setOnlineStatus() { // âŒ Not a component or custom Hook!
  const [onlineStatus, setOnlineStatus] = useOnlineStatus();
}
```
